python/ray/serve/_private/replica.py

```diff
@@ -814,145 +811,199 @@ async def call_reconfigure(self, user_config: Any):
-        `RayTaskError`.
+        Returns (request_args, asgi_args, receive_task).
+
+        The returned `receive_task` should be cancelled when the user method exits.
         """
-        logger.info(
-            f"Started executing request {request_metadata.request_id}",
-            extra={"log_to_stderr": False, "serve_access_log": True},
+        receive = ASGIReceiveProxy(
+            request_metadata.request_id,
+            request.http_proxy_handle,
+        )
+        receive_task = self._event_loop.create_task(receive.fetch_until_disconnect())
+        asgi_args = ASGIArgs(
+            scope=pickle.loads(request.pickled_asgi_scope),
+            receive=receive,
+            send=result_queue,
         )
+        if is_asgi_app:
+            request_args = asgi_args.to_args_tuple()
+        elif len(user_method_params) == 0:
+            # Edge case to support empty HTTP handlers: don't pass the Request
+            # argument if the callable has no parameters.
+            request_args = tuple()
+        else:
+            # Non-FastAPI HTTP handlers take only the starlette `Request`.
+            request_args = (asgi_args.to_starlette_request(),)
 
-        if request_metadata.is_http_request:
-            # For HTTP requests we always expect (scope, receive, send) as args.
-            assert len(request_args) == 3
-            scope, receive, send = request_args
+        return request_args, asgi_args, receive_task
 
-            if isinstance(self._callable, ASGIAppReplicaWrapper):
-                request_args = (scope, receive, send)
-            else:
-                request_args = (Request(scope, receive, send),)
-        elif request_metadata.is_grpc_request:
-            # Ensure the request args are a single gRPCRequest object.
-            assert len(request_args) == 1 and isinstance(request_args[0], gRPCRequest)
-            request_args = (pickle.loads(request_args[0].grpc_user_request),)
+    def _prepare_args_for_grpc_request(
+        self,
+        request: gRPCRequest,
+        request_metadata: RequestMetadata,
+        user_method_params: Dict[str, inspect.Parameter],
+    ) -> Tuple[Tuple[Any], Dict[str, Any]]:
+        """Prepare arguments for a user method handling a gRPC request.
 
-        runner_method = None
-        try:
-            runner_method = self._get_user_callable_method(request_metadata.call_method)
-            if inspect.isgeneratorfunction(runner_method) or inspect.isasyncgenfunction(
-                runner_method
-            ):
-                raise TypeError(
-                    f"Method '{runner_method.__name__}' is a generator function. "
-                    "You must use `handle.options(stream=True)` to call "
-                    "generators on a deployment."
-                )
+        Returns (request_args, request_kwargs).
+        """
+        request_args = (pickle.loads(request.grpc_user_request),)
+        if GRPC_CONTEXT_ARG_NAME in user_method_params:
```

Comment:
We'd also probably check the shape (type)

Reply:
this is existing code just moved